src/reducers/console.js

```diff
@@ -35,20 +62,38 @@ export default function console(stateIn, {type, payload, meta}) {
         ),
       );
     case 'EVALUATE_CONSOLE_ENTRY':
-      return payload.trim(' ') === '' ? state : state.set(
-        meta.key,
+      return payload.trim(' ') === '' ? state : state.setIn(
+        ['history', meta.key],
         new ConsoleEntry({expression: payload}),
-      );
+      ).
+        delete('currentInputValue').
+        delete('nextConsoleEntry').
+        delete('historyEntryIndex');
     case 'CLEAR_CONSOLE_ENTRIES':
       return initialState;
+    case 'CONSOLE_INPUT_CHANGED':
+      return state.set('currentInputValue', payload.value);
     case 'CONSOLE_LOG_PRODUCED':
-      return state.set(
-        meta.key,
+      return state.setIn(
+        ['history', meta.key],
         new ConsoleEntry({
           value: payload.value,
           evaluatedByCompiledProjectKey: payload.compiledProjectKey,
         }),
       );
+    case 'PREVIOUS_CONSOLE_HISTORY': {
+      const newHistoryEntryIndex = state.historyEntryIndex + 1;
```

Comment:
I don’t feel strongly about it, but you could inline these variables and remove the need for the weird case statement block form.